File: Python/ByRound/1839/1839_E_Decreasing_Game.py
# Problem: CF 1839 E - Decreasing Game
# https://codeforces.com/contest/1839/problem/E

"""
Algorithm: Game Theory + Sprague-Grundy Theorem + Nim-like Game
This problem is a variant of a Nim-like game where two players take turns decreasing two elements of an array simultaneously.
The key insight is to determine the winning player using the Sprague-Grundy theorem and nim-sum calculation.
We compute the nim-sum (XOR) of all array elements to determine who wins.
Time Complexity: O(n * sum(a)) where n is the length of array and sum(a) is the total sum of elements
Space Complexity: O(sum(a)) for the dp array storing nim-values
"""

import os
import sys

def f(u):
    # Function to send move to jury and receive response
    os.write(1, b"%d\n" % u)
    return int(input())


n = int(input())
a = [0] + list(map(int, input().split()))  # 1-indexed array
l = sum(a) + 5  # Upper bound for dp array size
dp = [0] * l
dp[0] = 1  # Base case: nim-sum 0 is losing for current player
s = 0  # Total sum of elements

# Build dp table using dynamic programming (similar to subset sum problem)
for i in range(1, n + 1):
    ai = a[i]
    # Traverse backwards to prevent using updated values in same iteration
    for j in range(s, -1, -1):
        if dp[j] and not dp[ai + j]:
            dp[ai + j] = i  # Mark which element contributes to this nim-sum
    s += ai  # Update total sum

# Determine winner based on nim-sum
# If total sum is odd or if we can't reach sum/2, first player wins
x = "First" if s % 2 or not dp[s // 2] else "Second"
os.write(1, b"%s\n" % x.encode())

if x == "First":
    # First player's turn: always choose the maximum element
    v = 1
    while v:
        u = a.index(max(a))  # Find index of maximum element
        v = f(u)  # Send move to jury and get opponent's move
        d = min(a[u], a[v])  # Calculate minimum of the two values
        a[u] -= d  # Decrease both elements
        a[v] -= d

else:
    # Second player's turn: determine winning strategy using dp
    c = [0] * (n + 1)  # Boolean array to identify winning moves
    z = s // 2  # Target nim-sum we want to avoid
    while z:
        c[dp[z]] = 1  # Mark elements involved in strategy
        z -= a[dp[z]]  # Backtrack to find components
    u = int(input())  # Receive first move from jury
    while u:
        # Find a move such that xor of current move and our strategy is zero
        for i in range(1, n + 1):
            if a[i] and c[i] ^ c[u]:
                v = i  # Valid move found
                break
        d = min(a[u], a[v])
        a[u] -= d
        a[v] -= d
        u = f(v)  # Send move and continue
exit()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1839/1839_E_Decreasing_Game.cpp
// Problem: CF 1839 E - Decreasing Game
// https://codeforces.com/contest/1839/problem/E

/*
 * Problem: Decreasing Game
 * 
 * Purpose: This code solves an interactive two-player game where players take turns 
 *          decreasing elements of an array. The goal is to determine whether the 
 *          first or second player has a winning strategy.
 * 
 * Algorithm: 
 *   - Use dynamic programming to compute all possible sums that can be achieved
 *     using subsets of the array elements.
 *   - If the total sum is odd, the first player has an advantage and wins.
 *   - If the total sum is even, we check if it's possible to split the array into two 
 *     subsets with equal sums. If so, we use a greedy strategy based on the subset 
 *     assignment to ensure a win.
 *   
 * Time Complexity: O(N * S), where N is the size of the array and S is the sum of all elements.
 * Space Complexity: O(N * S), for the dp table storing subset sums.
 * 
 * Techniques:
 *   - Dynamic Programming with bitsets for optimized subset sum calculation
 *   - Game theory strategy based on subset partitioning
 *   - Interactive problem handling with proper I/O flush
 */

#include <algorithm>
#include <bitset>
#include <functional>
#include <iostream>
#include <iterator>
#include <set>
#include <utility>
#include <vector>

using namespace std;
#define sz(x) ((int)(x).size())
typedef long long ll;
typedef pair<int, int> pii;
template <class T> void print(T &x) { cout << x; }
template <class T, class U> void print(pair<T, U> &p) {
  cout << "(";
  print(p.first);
  cout << ", ";
  print(p.second);
  cout << ")";
}
template <class T> void print(vector<T> &v) {
  cout << "{";
  if (sz(v))
    print(v[0]);
  for (int i = 1; i < sz(v); i++)
    cout << ", ", print(v[i]);
  cout << "}\n";
}
const int n = 301;
vector<bitset<n * n>> dp;
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(0);
  int N;
  cin >> N;
  vector<int> A(N);
  int sum = 0;
  for (int i = 0; i < N; i++) {
    cin >> A[i];
    sum += A[i];
  }
  dp = vector<bitset<n * n>>(N + 1);
  dp[0][0] = true;
  for (int i = 0; i < N; i++) {
    dp[i + 1] = dp[i] | (dp[i] << A[i]); // Update dp with new element
  }
  function<int(int)> f;
  bool isFirst = false;
  set<int> s1, s2;
  if (sum % 2 == 1 || !dp[N][sum / 2]) {
    // If sum is odd or no equal partition exists, first player wins by choosing any positive element
    f = [&](int other) {
      for (int i = 0; i < N; i++) {
        if (A[i] > 0)
          return i;
      }
      return -1;
    };
    isFirst = true;
  } else {
    // If sum is even and can be partitioned, strategy based on subset assignment
    int curV = sum / 2;
    for (int i = N - 1; i >= 0; i--) {
      if (dp[i][curV]) {
        s1.insert(i);
      } else {
        s2.insert(i);
        curV -= A[i];
      }
    }
    // Function to determine next move based on current player's role
    f = [&](int other) {
      set<int> *curS = &s1;
      if (s1.count(other)) // If other player's move is in s1, play in s2
        curS = &s2;
      for (int i : *curS) {
        if (A[i] > 0)
          return i;
      }
      return -1;
    };
  }
  cout << (isFirst ? "First" : "Second") << endl;
  int other = -3;
  while (true) {
    if (!isFirst) {
      cin >> other;
      other--;
    }
    if (other == -1)
      break;
    int me = f(other);
    cout << me + 1 << endl;
    if (isFirst) {
      cin >> other;
      other--;
    }
    if (other == -1)
      break;
    int minV = min(A[me], A[other]);
    A[me] -= minV;
    A[other] -= minV;
  }
}


// https://github.com/VaHiX/CodeForces/